### PR TITLE
[CSL-2375] Change the NodeInfo response to be more flexible.

### PR DESCRIPTION
**Pull request by parsonsmatt**
Huh, the template disappeared?

https://iohk.myjetbrains.com/youtrack/issue/CSL-2375

### CODE REVIEW

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 1]
```diff
@@ -1489,13 +1518,15 @@ instance ToSchema NodeInfo where
       & ("syncProgress"          --^ "Syncing progression, in percentage.")
       & ("blockchainHeight"      --^ "If known, the current blockchain height, in number of blocks.")
       & ("localBlockchainHeight" --^ "Local blockchain height, in number of blocks.")
-      & ("localTimeDifference"   --^ "Local time difference, in number of blocks.")
+      & ("localTimeInformation"  --^ "Information about the clock on this node.")
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: Whoops, sorry for that, looks like a wrong copy/paste from me :s ... 
Good catch :+1:

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 2]
```diff
@@ -1489,13 +1518,15 @@ instance ToSchema NodeInfo where
...
     )
 
 instance Arbitrary NodeInfo where
+    -- TODO(matt.parsons): The Swagger spec doesn't understand optional/maybe values for
+    -- some reason, so we create Just values instead.
+    --
+    -- This is fixed in this issue: https://github.com/GetShopTV/swagger2/issues/142
     arbitrary = NodeInfo <$> arbitrary
```
Comment (KtorZ) [inline wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: The issue seems to only concerns unary records, right ?

Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: Yes. We could add another field and work around it, but I'd rather see about bumping the swagger dependency.

Comment (KtorZ) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: But isn't the comment & the `fmap Just` redundant with the instance (and comment) of `TimeInfo` though?

Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: oh, that may be the case! i am investigating. good call :smile:

File: wallet-new/Makefile [hunk 3]
```diff
@@ -2,9 +2,12 @@ help: ## Print documentation
 	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
 
 ghcid: ## Run ghcid with the wallet-new project
-	ghcid -c "stack ghci cardano-sl-wallet-new"
+	ghcid \
+	    --command "stack ghci cardano-sl-wallet-new --ghci-options=-fno-code"
```
[the remark this answers is not in the document]
Comment (parsonsmatt) [reply wallet-new/Makefile]: it is *so* fast, ghcid is amazing

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 4]
```diff
@@ -144,7 +145,13 @@ type IsPropertiesMap m =
   (IxValue m ~ Referenced Schema, Index m ~ Text, At m, HasProperties Schema m)
 
 genericSchemaDroppingPrefix
-    :: forall a m proxy. (Generic a, ToJSON a, Arbitrary a, GToSchema (Rep a), IsPropertiesMap m)
+    :: forall a m proxy.
+    ( Generic a, ToJSON a, Arbitrary a, GToSchema (Rep a), IsPropertiesMap m
+    , GenericHasSimpleShape
+        a
+        "genericDeclareNamedSchemaUnrestricted"
+        (GenericShape (Rep a))
```
[the remark this answers is not in the document]
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: The Serokell fork of swagger was out of date. I think this instance means that less invalid types will be derived?

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs [hunk 5]
```diff
@@ -1277,12 +1274,17 @@ data NodeSettings = NodeSettings {
...
    } deriving (Show, Eq, Generic)
 
--- ORPHAN! TODO: Newtype this?
+-- See note [Version Orphan]
 instance ToSchema Version where
     declareNamedSchema _ =
         pure $ NamedSchema (Just "Version") $ mempty
             & type_ .~ SwaggerString
 
+-- Note [Version Orphan]
```
[the remark this answers is not in the document]
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: Good idea! `I` is meaningless on a codebase with this many contributors. Unless you're `blame`ing, anyway, but that can shift :smile:

File: wallet/src/Pos/Wallet/Web/Methods/Misc.hs [hunk 6]
```diff
@@ -176,14 +176,14 @@ syncProgress = do
...
   where
-    diff :: NtpStatus -> Integer
+    diff :: NtpStatus -> Maybe Integer
     diff = \case
-        NtpDrift time -> toMicroseconds time
-        NtpSyncPending -> 0
-        NtpSyncUnavailable -> 0
+        NtpDrift time -> Just (toMicroseconds time)
+        NtpSyncPending -> Nothing
+        NtpSyncUnavailable -> Nothing
```
[the remark this answers is not in the document]
Comment (parsonsmatt) [reply wallet/src/Pos/Wallet/Web/Methods/Misc.hs]: We can easily include another field on the object, too -- "if the `drift` is null, then check the `reason` for why"

File: wallet-new/src/Cardano/Wallet/API/V1/Types.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (adinapoli-iohk) [reply wallet-new/src/Cardano/Wallet/API/V1/Types.hs]: One can only hope! 😁

File: wallet/src/Pos/Wallet/Web/Methods/Misc.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (adinapoli-iohk) [reply wallet/src/Pos/Wallet/Web/Methods/Misc.hs]: Yup, that would also works 😉

**Pull request merged**